src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -545,37 +547,115 @@ where
             Poll::Pending => source.block_until_ready()?,
         }
     };
-    match deps.iter().map(|p| p.package_id()).max() {
-        Some(pkgid) => {
-            let pkg = Box::new(source).download_now(pkgid, config)?;
-            Ok(pkg)
+    if let Some(pkgid) = deps.iter().map(|p| p.package_id()).max() {
+        let pkg = Box::new(source).download_now(pkgid, config)?;
+        Ok(pkg)
+    } else {
+        let spe = SelectPackageError::new(dep, source);
```

Comment:
When extracting code into a function, please split that out into a separate commit so its easier to review the two halves rather than having to compare disparate code.

Reply:
You are right, i didn't structure the PR as well as I could have